### ISSUE TITLE
[BUG]: "ConvertFrom-Json : Invalid JSON primitive: ." in "Disable Visual Studio Code Data Collection"

### ISSUE THREAD

**Issue by ItsIgnacioPortal** [undergroundwires/privacy.sexy#196]
### Description

I ran the generated privacy-script.bat by double clicking on it, and I got the following errors:
```powershell
--- Disable Visual Studio Code telemetry
ConvertFrom-Json : Invalid JSON primitive: .
At line:1 char:216
+ ... $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json  ...
+                                                  ~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [ConvertFrom-Json], ArgumentException
    + FullyQualifiedErrorId : System.ArgumentException,Microsoft.PowerShell.Commands.ConvertFromJsonCommand

Add-Member : Cannot bind argument to parameter 'InputObject' because it is null.
At line:1 char:242
+ ... on; $json | Add-Member -Type NoteProperty -Name 'telemetry.enableTele ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Add-Member], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.AddMemberCo
   mmand

--- Disable Visual Studio Code crash reporting
ConvertFrom-Json : Invalid JSON primitive: .
At line:1 char:216
+ ... $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json  ...
+                                                  ~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [ConvertFrom-Json], ArgumentException
    + FullyQualifiedErrorId : System.ArgumentException,Microsoft.PowerShell.Commands.ConvertFromJsonCommand

Add-Member : Cannot bind argument to parameter 'InputObject' because it is null.
At line:1 char:242
+ ... on; $json | Add-Member -Type NoteProperty -Name 'telemetry.enableCras ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Add-Member], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.AddMemberCo
   mmand

--- Do not run Microsoft online experiments
ConvertFrom-Json : Invalid JSON primitive: .
At line:1 char:216
+ ... $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json  ...
+                                                  ~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [ConvertFrom-Json], ArgumentException
    + FullyQualifiedErrorId : System.ArgumentException,Microsoft.PowerShell.Commands.ConvertFromJsonCommand

Add-Member : Cannot bind argument to parameter 'InputObject' because it is null.
At line:1 char:242
+ ... on; $json | Add-Member -Type NoteProperty -Name 'workbench.enableExpe ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Add-Member], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.AddMemberCo
   mmand

--- Choose manual updates over automatic updates
ConvertFrom-Json : Invalid JSON primitive: .
At line:1 char:216
+ ... $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json  ...
+                                                  ~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [ConvertFrom-Json], ArgumentException
    + FullyQualifiedErrorId : System.ArgumentException,Microsoft.PowerShell.Commands.ConvertFromJsonCommand

Add-Member : Cannot bind argument to parameter 'InputObject' because it is null.
At line:1 char:242
+ ... on; $json | Add-Member -Type NoteProperty -Name 'update.mode' -Value  ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Add-Member], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.AddMemberCo
   mmand

--- Show Release Notes from Microsoft online service after an update
ConvertFrom-Json : Invalid JSON primitive: .
At line:1 char:216
+ ... $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json  ...
+                                                  ~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [ConvertFrom-Json], ArgumentException
    + FullyQualifiedErrorId : System.ArgumentException,Microsoft.PowerShell.Commands.ConvertFromJsonCommand

Add-Member : Cannot bind argument to parameter 'InputObject' because it is null.
At line:1 char:242
+ ... on; $json | Add-Member -Type NoteProperty -Name 'update.showReleaseNo ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Add-Member], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.AddMemberCo
   mmand

--- Automatically check extensions from Microsoft online service
ConvertFrom-Json : Invalid JSON primitive: .
At line:1 char:216
+ ... $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json  ...
+                                                  ~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [ConvertFrom-Json], ArgumentException
    + FullyQualifiedErrorId : System.ArgumentException,Microsoft.PowerShell.Commands.ConvertFromJsonCommand

Add-Member : Cannot bind argument to parameter 'InputObject' because it is null.
At line:1 char:242
+ ... on; $json | Add-Member -Type NoteProperty -Name 'extensions.autoCheck ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Add-Member], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.AddMemberCo
   mmand

--- Fetch recommendations from Microsoft only on demand
ConvertFrom-Json : Invalid JSON primitive: .
At line:1 char:216
+ ... $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json  ...
+                                                  ~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [ConvertFrom-Json], ArgumentException
    + FullyQualifiedErrorId : System.ArgumentException,Microsoft.PowerShell.Commands.ConvertFromJsonCommand

Add-Member : Cannot bind argument to parameter 'InputObject' because it is null.
At line:1 char:242
+ ... on; $json | Add-Member -Type NoteProperty -Name 'extensions.showRecom ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Add-Member], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.AddMemberCo
   mmand

--- Automatically fetch git commits from remote repository
ConvertFrom-Json : Invalid JSON primitive: .
At line:1 char:216
+ ... $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json  ...
+                                                  ~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [ConvertFrom-Json], ArgumentException
    + FullyQualifiedErrorId : System.ArgumentException,Microsoft.PowerShell.Commands.ConvertFromJsonCommand

Add-Member : Cannot bind argument to parameter 'InputObject' because it is null.
At line:1 char:242
+ ... on; $json | Add-Member -Type NoteProperty -Name 'git.autofetch' -Valu ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Add-Member], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.AddMemberCo
   mmand

--- Fetch package information from NPM and Bower
ConvertFrom-Json : Invalid JSON primitive: .
At line:1 char:216
+ ... $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json  ...
+                                                  ~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [ConvertFrom-Json], ArgumentException
    + FullyQualifiedErrorId : System.ArgumentException,Microsoft.PowerShell.Commands.ConvertFromJsonCommand

Add-Member : Cannot bind argument to parameter 'InputObject' because it is null.
At line:1 char:242
+ ... on; $json | Add-Member -Type NoteProperty -Name 'npm.fetchOnlinePacka ...
+                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidData: (:) [Add-Member], ParameterBindingValidationException
    + FullyQualifiedErrorId : ParameterArgumentValidationErrorNullNotAllowed,Microsoft.PowerShell.Commands.AddMemberCo
   mmand

Press any key to continue . . .
```

### OS

Windows Enterprise 22H2

### Reproduction steps

1. Go to privacy.sexy
2. Search "visual studio code"
3. Select "Disable visual studio code data collection"
4. Download the generated script
5. Run it

### Scripts

```
@echo off
:: https://privacy.sexy ΓÇö v0.11.4 ΓÇö Tue, 24 Jan 2023 00:24:27 GMT
:: Ensure admin privileges
fltmc >nul 2>&1 || (
    echo Administrator privileges are required.
    PowerShell Start -Verb RunAs '%0' 2> nul || (
        echo Right-click on the script and select "Run as administrator".
        pause & exit 1
    )
    exit 0
)


:: ----------------------------------------------------------
:: -----------Disable Visual Studio Code telemetry-----------
:: ----------------------------------------------------------
echo --- Disable Visual Studio Code telemetry
PowerShell -ExecutionPolicy Unrestricted -Command "$jsonfile = "^""$env:APPDATA\Code\User\settings.json"^""; if (!(Test-Path $jsonfile -PathType Leaf)) {; Write-Host "^""No updates. Settings file was not at $jsonfile"^""; exit 0; }; $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json | Add-Member -Type NoteProperty -Name 'telemetry.enableTelemetry' -Value $false -Force; $json | ConvertTo-Json | Set-Content $jsonfile"
:: ----------------------------------------------------------


:: ----------------------------------------------------------
:: --------Disable Visual Studio Code crash reporting--------
:: ----------------------------------------------------------
echo --- Disable Visual Studio Code crash reporting
PowerShell -ExecutionPolicy Unrestricted -Command "$jsonfile = "^""$env:APPDATA\Code\User\settings.json"^""; if (!(Test-Path $jsonfile -PathType Leaf)) {; Write-Host "^""No updates. Settings file was not at $jsonfile"^""; exit 0; }; $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json | Add-Member -Type NoteProperty -Name 'telemetry.enableCrashReporter' -Value $false -Force; $json | ConvertTo-Json | Set-Content $jsonfile"
:: ----------------------------------------------------------


:: ----------------------------------------------------------
:: ---------Do not run Microsoft online experiments----------
:: ----------------------------------------------------------
echo --- Do not run Microsoft online experiments
PowerShell -ExecutionPolicy Unrestricted -Command "$jsonfile = "^""$env:APPDATA\Code\User\settings.json"^""; if (!(Test-Path $jsonfile -PathType Leaf)) {; Write-Host "^""No updates. Settings file was not at $jsonfile"^""; exit 0; }; $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json | Add-Member -Type NoteProperty -Name 'workbench.enableExperiments' -Value $false -Force; $json | ConvertTo-Json | Set-Content $jsonfile"
:: ----------------------------------------------------------


:: ----------------------------------------------------------
:: -------Choose manual updates over automatic updates-------
:: ----------------------------------------------------------
echo --- Choose manual updates over automatic updates
PowerShell -ExecutionPolicy Unrestricted -Command "$jsonfile = "^""$env:APPDATA\Code\User\settings.json"^""; if (!(Test-Path $jsonfile -PathType Leaf)) {; Write-Host "^""No updates. Settings file was not at $jsonfile"^""; exit 0; }; $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json | Add-Member -Type NoteProperty -Name 'update.mode' -Value 'manual' -Force; $json | ConvertTo-Json | Set-Content $jsonfile"
:: ----------------------------------------------------------


:: Show Release Notes from Microsoft online service after an update
echo --- Show Release Notes from Microsoft online service after an update
PowerShell -ExecutionPolicy Unrestricted -Command "$jsonfile = "^""$env:APPDATA\Code\User\settings.json"^""; if (!(Test-Path $jsonfile -PathType Leaf)) {; Write-Host "^""No updates. Settings file was not at $jsonfile"^""; exit 0; }; $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json | Add-Member -Type NoteProperty -Name 'update.showReleaseNotes' -Value $false -Force; $json | ConvertTo-Json | Set-Content $jsonfile"
:: ----------------------------------------------------------


:: Automatically check extensions from Microsoft online service
echo --- Automatically check extensions from Microsoft online service
PowerShell -ExecutionPolicy Unrestricted -Command "$jsonfile = "^""$env:APPDATA\Code\User\settings.json"^""; if (!(Test-Path $jsonfile -PathType Leaf)) {; Write-Host "^""No updates. Settings file was not at $jsonfile"^""; exit 0; }; $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json | Add-Member -Type NoteProperty -Name 'extensions.autoCheckUpdates' -Value $false -Force; $json | ConvertTo-Json | Set-Content $jsonfile"
:: ----------------------------------------------------------


:: ----------------------------------------------------------
:: ---Fetch recommendations from Microsoft only on demand----
:: ----------------------------------------------------------
echo --- Fetch recommendations from Microsoft only on demand
PowerShell -ExecutionPolicy Unrestricted -Command "$jsonfile = "^""$env:APPDATA\Code\User\settings.json"^""; if (!(Test-Path $jsonfile -PathType Leaf)) {; Write-Host "^""No updates. Settings file was not at $jsonfile"^""; exit 0; }; $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json | Add-Member -Type NoteProperty -Name 'extensions.showRecommendationsOnlyOnDemand' -Value $true -Force; $json | ConvertTo-Json | Set-Content $jsonfile"
:: ----------------------------------------------------------


:: ----------------------------------------------------------
:: --Automatically fetch git commits from remote repository--
:: ----------------------------------------------------------
echo --- Automatically fetch git commits from remote repository
PowerShell -ExecutionPolicy Unrestricted -Command "$jsonfile = "^""$env:APPDATA\Code\User\settings.json"^""; if (!(Test-Path $jsonfile -PathType Leaf)) {; Write-Host "^""No updates. Settings file was not at $jsonfile"^""; exit 0; }; $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json | Add-Member -Type NoteProperty -Name 'git.autofetch' -Value $false -Force; $json | ConvertTo-Json | Set-Content $jsonfile"
:: ----------------------------------------------------------


:: ----------------------------------------------------------
:: -------Fetch package information from NPM and Bower-------
:: ----------------------------------------------------------
echo --- Fetch package information from NPM and Bower
PowerShell -ExecutionPolicy Unrestricted -Command "$jsonfile = "^""$env:APPDATA\Code\User\settings.json"^""; if (!(Test-Path $jsonfile -PathType Leaf)) {; Write-Host "^""No updates. Settings file was not at $jsonfile"^""; exit 0; }; $json = Get-Content $jsonfile | Out-String | ConvertFrom-Json; $json | Add-Member -Type NoteProperty -Name 'npm.fetchOnlinePackageInfo' -Value $false -Force; $json | ConvertTo-Json | Set-Content $jsonfile"
:: ----------------------------------------------------------


pause
exit /b 0
```

### Additional information

Visual Studio Code v1.74.3 is installed

**Comment by undergroundwires** [comment]
Hi @ItsIgnacioPortal , I could not reproduce this. It works fine for me (tested on both Windows 10 and 11). The only reason I can think is that your `settings.json` is corrupted. This may happen:

- If the `settings.json` file is empty or contains invalid JSON, `ConvertFrom-Json` will not be able to parse the content.
- If for any reason `Get-Content` fails to read the content of the file. This can happen if the file is locked by another process or there are insufficient permissions to read the file.

It is most likely that your `settings.json` was empty (had no lines) and therefore invalid JSON.

I do not consider this bug of privacy.sexy, but more like improvement, but in any case I will improve the script to handle these edge cases:

- Can handle if the `settings.json` is empty (I will add this to Linux VSCode modifications too to keep them consistent).
- Show better errors if JSON file is corrupted.
- Show better errors if JSON file cannot be read.

So I will add these improvements for this issue and increase overall robustness:

- Add a check to exit the script with an error message if the attempt to parse the JSON content fails.
- Omit the `OutString` cmdlet from the pipeline in the script for converting JSON file content to a PowerShell object. `Out-String` is unnecessary in this context because `Get-Content` already outputs the file content as a string array, which `ConvertFrom-Json` effectively. Additionally, using `Out-String` could potentially introduce issues concatenating file content into a single string, causing `ConvertFrom-Json` to fail when processing pretty-printed JSON. By removing `Out-String`, the script is streamlined and potential errors are avoided.
- Add logic to handle empty settings file. Add an additional check for empty settings file, if the file is empty, the script writes a default empty JSON object (`{}`) to the file. The operation is logged to ensure transparency, notifying the user of the action taken. This change removes fails due to empty setting files.
- When reverting, do not fail if the setting file is missing because it means that default settings are already in-place.
- When reverting, show informative message if the key does not exist or does not have the value set by privacy.sexy and do not take any further action.

**Comment by undergroundwires** [comment]
The fix/improvement fix is released in [0.12.5](https://github.com/undergroundwires/privacy.sexy/releases/tag/0.12.5) 🚀

**Comment by ItsIgnacioPortal** [comment]
Awesome! 🔥🔥
